### DISCUSSION TITLE
remove ujson dependency (it is dead)

### DISCUSSION THREAD

**Discussion by hnykda** [fastapi/fastapi#9155] [Questions] [answered]
FastAPI is using `ujson` which has not been actively developed for about 3 years by now. [Here is a relevant issue](https://github.com/esnme/ultrajson/issues/343). It's starting to be problematic to actually compile `ujson` on newer platforms.

Possible alternative maybe `orjson`.

**Comment by dmontagu** [comment]
I would be in favor of recommending `orjson` instead of `ujson`.

I recall the creator of `orjson` reaching about this many months ago and there was some confusion due to slow responses, and also some issues on the pydantic side, but I think many/most of the barriers to usage have since been removed. If `orjson` seems to be taking over as the go-to high-performance json library, I think it would make sense to use that for development and recommend it.

To the extent that there is any difference in json encoding/decoding between json and orjson (or ujson and orjson), it would probably be best to document it.

I'd be willing to review a PR making this change (including updates to the docs). (I'd leave the ultimate decision about merging to @tiangolo though.)

**Comment by tiangolo** [comment, marked as answer]
Yeah, I've been intending to do it for quite some time now... :sweat_smile: 

In fact, you can sort of do it right now (and since several versions ago). 

---

Here's my plan:

I want to export everything from Starlette, to avoid confusions of the type "do I import this from FastAPI or from Starlette?", I've seen reports of that and even I've struggled with that. For example, FastAPI has its own `HTTPException`.

With that, I want to have a response `ORJSONResponse` exported from `fastapi.responses`, among Starlette's responses.

In fact, I added tests for creating custom responses (as part of a PR), testing specifically creating orjson responses: https://github.com/tiangolo/fastapi/blob/f803c77515662cd2382674d9fc8df1e6641b3ba7/tests/test_default_response_class.py#L9-L13
And you can replicate creating that response class and using it. It's just 3 lines (you would remove the custom `media_type`), but still, it should be provided by FastAPI (among all the other responses from Starlette).

```Python
app = FastAPI(default_response_class=ORJSONResponse)
```

---

Also, as a side note, UJSON is not really used by default anywhere. It's optional, and you would have to use the `UJSONResponse` class for it to be used, the same way you would use the (new/custom/future) `ORJSONResponse`.

**Comment by hnykda** [comment]
Well, that's elegant sir :bowing_man: :clap: .

> Also, as a side note, UJSON is not really used by default anywhere. It's optional, and you would have to use the UJSONResponse class for it to be used, the same way you would use the (new/custom/future) ORJSONResponse.

Then I don't understand why it's in [requirements](https://github.com/tiangolo/fastapi/blob/master/pyproject.toml#L81). It's [written](https://github.com/tiangolo/fastapi#optional-dependencies) that it's an optional dependency for Pydantic, but it seems that Pydantic has already moved away from ujson [here](https://github.com/samuelcolvin/pydantic/issues/589). Seems that their [requirements](https://github.com/samuelcolvin/pydantic/blob/master/requirements.txt) nor [setup.py](https://github.com/samuelcolvin/pydantic/blob/master/setup.py) don't have it anymore.

**Comment by dmontagu** [comment]
@hnykda the “all” set of requirements there includes everything necessary to make use of *all documented features*, including those that are optional. It is intended more for development/reference purposes than for use by consumers.

**Comment by hnykda** [comment]
Ah, got it! Thanks. I believe this can be closed then.

Personally, I would not reference any obsolete dependencies in my project even if they are optional (especially "future-looking" like Python 3.6+, latest libs, ...) . Reasons for that: helping the community to get rid dangerous/old/unmaintained libs and to avoid possibly repealing people who do not like to install stuff which are referencing such unmaintained projects.

**Comment by dagnelies** [comment]
Apparently, `ujson` found a new maintainer and is now alive and kicking again. Well, dunno about the kicking part, but releases are pretty fresh: 

https://pypi.org/project/ujson/

and repo seems active again. Just mentioning this. Dunno which one is the better lib.

**Comment by dagnelies** [comment]
Nevermind, apparently `orjson` is a fine choice too. Apparently one of the fastest.

**Comment by ahmadazizi** [comment]
Just to mention, installing fastapi on windows platform (python 3.9) fails, since orjson neither provide the compatible wheel nor could be compiled.

**Comment by jacopofar** [comment]
[On Debian stable installing orjson currently fails](https://github.com/ijl/orjson/issues/136) because it targets `manylinux2014` which is supported by pip since version `19.3`, but Debian stable install pip `18.1` (if installed through Debian repos `python3-pip`).

In this case running `python3 -m pip install --upgrade pip` works, but the error can be tricky to detect

**Comment by ahtik** [comment]
@ahmadazizi,  `orjson` wheels are now available also for Windows Python 3.9.

**Comment by ycd** [comment]
Since Starlette dropped support for `UJSONResponse` in https://github.com/encode/starlette/pull/1047 (Requires Starlette 0.14.1), we can remove it too, `ujson` causes build fails on many distributions due C Dependencies (See #2811 ). 

This require changes in #2335, since it re-adds `UJSONResponse`.

**Comment by omidmaldar** [comment]
Regarding ujson, there is a [security alert](https://github.com/advisories/GHSA-fh56-85cw-5pq6) reported with moderate severity which is patched in 5.1.0.  Fastapi 0.74.0 [requires](https://github.com/tiangolo/fastapi/blob/1ce16c2f40a75a4d97233bf4e331ed563eda47cf/pyproject.toml#L93) `ujson (>=4.0.1,<5.0.0)` which is not safe.

**Comment by melt7777** [comment]
Hello all, I did find a workaround which allows us to build images without the ujson. I am using the `python-poetry` package manager, instead of requirements.txt. 

FastAPI is fantastic so far, thank you all for your hard work on this project. 

Here, roughly the steps I used:

1. `poetry init` 
2. `poetry add fastapi -E all` - it installed the ujson package and placed the ujson bits into the pyproject.toml and the newly created poetry.lock file. 
3. Manually edit the .toml and .lock files and remove all the ujson items.
4. Build image as usual and it will not have the ujson or its vuln in the image. 

Hope this helps someone until they scan make the permanent fix.

Thank you,

~ melt

**Comment by alexted** [comment]
Are you sure the **ujson** is EOL? According to [its repository](https://github.com/ultrajson/ultrajson), the last release was in mid-September, and the last commit was 3 weeks ago - the project seems to be quite alive.
Another point is that it is inferior to **orjson** in performance. In that case, it can be left as an option.

**Comment by hofrob** [comment]
Just as an update. This is what the pypi page says:

> [!WARNING] UltraJSON's architecture is fundamentally ill-suited to making changes without risk of introducing new security vulnerabilities. As a result, this library has been put into a maintenance-only mode. Support for new Python versions will be added and critical bugs and security issues will still be fixed but all other changes will be rejected. Users are encouraged to migrate to [orjson](https://pypi.org/project/orjson/) which is both much faster and less likely to introduce a surprise buffer overflow vulnerability in the future.

Additionally, I just want to mention that in my project, `ujson` was added as a dependency after upgrading fastapi to [0.111.0](https://github.com/tiangolo/fastapi/releases/tag/0.111.0). Is that expected? Does this have something to do with the refactoring for `fastapi-slim` and `fastapi-slim[standard]`?

**Reply by hofrob** [reply]
> Does this have something to do with the refactoring for fastapi-slim and fastapi-slim[standard]?

Yes, it does. I had to replace `fastapi` as a dependency with `fastapi-slim`.
